core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -418,6 +425,64 @@ public final SegmentsStats segmentsStats() {
                 stats.addTermVectorsMemoryInBytes(guardedRamBytesUsed(segmentReader.getTermVectorsReader()));
                 stats.addNormsMemoryInBytes(guardedRamBytesUsed(segmentReader.getNormsReader()));
                 stats.addDocValuesMemoryInBytes(guardedRamBytesUsed(segmentReader.getDocValuesReader()));
+
+                if (!detailed) {
+                    // Gathering index file stats may be costly. Don't do it unless specified
+                    continue;
+                }
+
+                Directory directory = null;
+                SegmentCommitInfo segmentCommitInfo = segmentReader.getSegmentInfo();
+                boolean useCompoundFile = segmentCommitInfo.info.getUseCompoundFile();
+                if (useCompoundFile) {
+                    try {
+                        directory = engineConfig.getCodec().compoundFormat().getCompoundReader(segmentReader.directory(), segmentCommitInfo.info, IOContext.READ);
+                    } catch (IOException e) {
+                        logger.warn("Error when opening compound reader for Directory [{}] and SegmentCommitInfo [{}]", e,
+                                    segmentReader.directory(), segmentCommitInfo);
+                    }
+                } else {
+                    directory = segmentReader.directory();
+                }
+
+                // Couldn't open compound reader, skip
+                if (directory == null) {
+                    continue;
+                }
+
+                String[] files = new String[0];
+                try {
+                    files = directory.listAll();
+                } catch (IOException e) {
+                    logger.warn("Couldn't list Directory [{}]", e, directory);
+                }
+
+                ImmutableOpenMap.Builder<String, Long> map = ImmutableOpenMap.builder();
+                for (String file : files) {
+                    String extension = IndexFileNames.getExtension(file);
+                    if (extension == null) {
+                        continue;
```

Comment:
Hmm did you see a `null` extension?  Can you add a comment explaining?  It's spooky if this happens because it becomes a loop-hole in the stats: what if Lucene adds a gigantic per-segment file with no extension :)